### DISCUSSION TITLE
Function w/ Service Bus Trigger not Scaling down

### DISCUSSION THREAD

**Discussion by JFolberth** [Azure/azure-functions-durable-extension#2036] [Q&A] [answered]
We have a function which is being triggered off a Service Bus. There is a steady stream of messages coming into the queue; however, historically never enough to need more than one instance of the function running.

A few months ago, our function app scaled out unexpectedly to ~60 instances and never scaled back in.

We placed a max scale limit and it happened again.

We'd like to know if there is a way to mitigate this from happening. We do not believe setting a max scale of 2 is appropriate as this would be counter intuitive to having consumption-based function apps.  Does anyone have any additional ideas or experience with this issue?

**Comment by olitomlinson** [comment]
@JFolberth just for clarity, are you using Durable Functions in your project?

**Reply by JFolberth** [reply]
Yes we are.

**Reply by olitomlinson** [reply]
@JFolberth thanks

Were there any significant environmental/code changes at the time of the odd behaviour starting? Did you deploy a specific version of code? Upgrade any dependencies/nuget packages? Did you change anything in `host.json`? did you upgrade .NET version or the Azure functions runtime version? etc

I personally find it useful to use App Insights and then drill into the telemetry around the time of the change. Use graphs as a visual aid to see if you can identify any changes or patterns.

If none of that helps, one of the team may be able to look at some internal telemetry (if it goes back that far!) with a view to understand what was causing the constant scale-out of your App. If you could provide the following details that would be great :

### App Details

- **Durable Functions extension version (e.g. v1.8.3)**:
- **Azure Functions runtime version (1.0 / 2.0 / 3.0 etc)**:
- **Programming language used**:

### Screenshots
> If applicable, add screenshots to help explain your problem.

### If deployed to Azure
> We have access to a lot of telemetry that can help with investigations. Please provide as much of the following information as you can to help us investigate!

- **Timeframe issue observed**:
- **Function App name**:
- **Function name(s)**:
- **Azure region**:
- **Orchestration instance ID(s)**:
- **Azure storage account name**:

**Reply by davidmrdavid** [reply]
Thanks @olitomlinson. Indeed, we'll need the information in the template above in order to begin and investigation. If the issue occurred relatively recently, then we should still have telemetry to use for diagnosis :) 

I can't think on top of my head as to what kind of behavior would cause a random scale-out, so I suspect there would have been some external change (traffic, code-wise, config-wise) as it's been suggested. In any case, we can investigate once we have the information requested above ^. Thanks!

**Reply by JFolberth** [reply]
@olitomlinson @davidmrdavid nothing was changed with the application nor recent deployments nor configuration. It randomly scaled out from our perspective.

We opened case 2108120040003836 with support after the initial issue a few months ago and struggled with getting the proper attention for quite a while.

We did ultimately get a response and was advised to share here for additional information or things we could or others could to mitigate this from happening.

I believe the case notes will contain all the requested info.

**Reply by davidmrdavid** [reply]
I'm sorry that your case didn't receive proper attention, @JFolberth. I'm reaching out internally to try to get access to this case. However, we might be a little slowed down due to folks being out of office near the end of year holidays.

In the meantime, it would speed things up if you could fill the issue template shared above, but I'll look to see if I can access the case before you get to it. I'll update this thread if so.

**Reply by davidmrdavid** [reply]
It seems I can review your internal case now, I'm reading it now

**Reply by JFolberth** [reply]
@davidmrdavid I am out myself so don't have the information on hand. I will get it to you when I can. I do know the initial logs were not retrieved in time by MS. However, MS was able to capture the logs the second time the function scaled out unexpectedly..

**Reply by davidmrdavid** [reply]
@JFolberth, unfortunately, it seems that we no longer have telemetry for the original date of the incident, as it occurred about half a year ago from what I gather.

Has the issue occurred again since? Maybe I can try with a more recent timestamp.

That said, I did take a look at your app configuration and I couldn't find any Durable Functions bindings when looking at data for the last 3 days. Can you share what specific Function Names use Durable Functions? In particular, what Function name is your orchestrator and/or Entity. I just want to make validate that this is Durable Functions-related.

On a related note, I noticed that the config variable `FUNCTIONS_WORKER_RUNTIME` isn't set, which can cause issues. Please set it as per [these](https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings#functions_worker_runtime) instructions. Thanks!

**Edit:** I just noticed your last comment. No worries, at your earliest convenience, please provide me with the date of the second time (or ideally the most recent time) this happened. I'll check-in internally for those captured logs

**Reply by JFolberth** [reply]
Hey @davidmrdavid apologies for being a little unresponsive to this one. The event last occurred on 10/10.

**Reply by davidmrdavid** [reply]
Thanks @JFolberth, for the ping. I also just returned from vacation so my apologies for the delayed response.
@cgillum: before diving too deeply, have you seen the behavior described here before? Do any theories come to mind?

**Reply by cgillum** [reply]
This sounds more like an issue with the Consumption plan scaling infra rather than Durable Functions. If the maximum instance count setting is not being honored, then it's best to open a support ticket to have the Azure Functions scale team have a look. It would be good to also mention that scaling in is not happening. This is unfortunately an issue that many users have seen when there's a steady stream of Service Bus messages.

**Reply by JFolberth** [reply]
@cgillum  The function app is stopping at the max; however, if this is not setup it will scale out more than is required and never scale back in. (Talking 12 instances when there are clearly no messages in the queue type scenario) We did test with a non-durable function and got the same result when there is a Service Bus trigger w/ a steady stream of messages.

Posted on here based on your feedback on any potential mitigation or work arounds the engineering team may have when the function is "over-scaled".

**Reply by cgillum** [reply]
@JFolberth, you mentioned that there are no messages in the queue, which is interesting. A long time back we did see issues where apps would fail to scale in if there were _deferred_ messages in the Service Bus queues, even though there weren't any receivable messages. Could that be the issue you're running into?

**Comment by mluker** [comment, marked as answer]
I wanted to add a little more context to this as I was part of the original issue/troubleshooting and working directly with @JFolberth.

A change I made that seemed to produce more reliable scaling for our function was to grant 'Manage' rights for the trigger on the service bus. This stopped the function from immediately scaling out to ~65 instances. Before making that change, the scale out would happen even with with no messages or trigger executions. It would just scale out for no good reason and never scale back in. 

It ran well for awhile, maybe a few weeks, without scaling but then we had a scale out to ~10 instances (I had set the scale to max at 10 just incase) with basically no traffic and it didn't scale back in.

Based on MS docs Listen rights should be enough, Manage rights will only cause difference in the sensitivity of scale controller's scale out/in behavior which is mentioned to be "more efficient" in official document. The Manage rights are supposed to just bring in one more scale in/out rule to the equation. From my experience, this is not the case. Without Manage, it will scale out and not back in.

**Reply by cgillum** [reply]
@mluker that's a really interesting observation. Would you mind opening an Azure Functions support request and sharing this info so that the platform engineering team can take a look? It's possible there may be some regression that went into the platform recently.

**Reply by JFolberth** [reply]
@cgillum Would it be alright to leverage the existing ticket? I believe there is some history and communication of what Matt has explained in the ticket as well as additional supporting documentation.

**Reply by cgillum** [reply]
@JFolberth yes, if you have a ticket already, that should be fine.

**Reply by JFolberth** [reply]
@cgillum Ticket 2108120040003836. The tech currently assigned has been helpful and aligned that there are some anomalies going on that we haven't been able to explain.

**Reply by davidmrdavid** [reply]
Great! I just want to re-emphasize that, from what I understand about the symptoms and @cgillum's assessment so far, this doesn't seem to be something the Durable Extension would manage. That's to say that we might not be the right team to diagnose this issue.

@JFolberth, my recommendation would be to continue working with your support engineer for the time being, especially if the current assigned person is moving the investigation forward. They  also have a means to reach out to the appropriate product group if that's required for your case. Finally, a big part of the Functions ecosystem is open source, so if the support engineer can identify the faulty component, there is likely a public repo we can use to reach out to the appropriate team; but we have to get to that point first.
